File: model-client/src/main/java/org/modelix/model/persistent/HashUtil.java
package org.modelix.model.persistent;



import java.util.regex.Pattern;
import java.nio.charset.Charset;
import java.nio.charset.StandardCharsets;
import java.security.MessageDigest;
import java.util.Base64;
import java.security.NoSuchAlgorithmException;
import java.util.Iterator;
import java.util.regex.Matcher;

public class HashUtil {

  private static final Pattern HASH_PATTERN = Pattern.compile("[a-zA-Z0-9\\-_]{43}");
  private static final Charset UTF8 = StandardCharsets.UTF_8;

  public static String sha256(byte[] input) {
    try {
      MessageDigest digest = MessageDigest.getInstance("SHA-256");
      digest.update(input);
      byte[] sha256Bytes = digest.digest();
      return Base64.getUrlEncoder().withoutPadding().encodeToString(sha256Bytes);
    } catch (NoSuchAlgorithmException ex) {
      throw new RuntimeException(ex);
    }
  }

  public static String sha256(String input) {
    return sha256(input.getBytes(UTF8));
  }

  public static boolean isSha256(String value) {
    if (value == null) {
      return false;
    }
    if (value.length() != 43) {
      return false;
    }
    return HASH_PATTERN.matcher(value).matches();
  }

  public static Iterable<String> extractSha256(final String input) {
    return new Iterable<String>() {
      @Override
      public Iterator<String> iterator() {
        return new Iterator<String>() {
          private Matcher matcher = HASH_PATTERN.matcher(input);
          private boolean hasNext;
          private boolean hasNextInitialized;
          public void ensureInitialized() {
            if (!(hasNextInitialized)) {
              hasNext = matcher.find();
              hasNextInitialized = true;
            }
          }
          @Override
          public boolean hasNext() {
            ensureInitialized();
            return hasNext;
          }

          @Override
          public String next() {
            ensureInitialized();
            hasNextInitialized = false;
            return matcher.group();
          }
        };
      }
    };
  }

  public static String base64encode(String input) {
    return Base64.getUrlEncoder().withoutPadding().encodeToString(input.getBytes(UTF8));
  }

  public static String base64decode(String input) {
    return new String(Base64.getUrlDecoder().decode(input.getBytes(UTF8)), UTF8);
  }

}

File: model-client/src/main/java/org/modelix/model/persistent/HashUtil.kt
package org.modelix.model.persistent

import java.nio.charset.StandardCharsets
import java.security.MessageDigest
import java.security.NoSuchAlgorithmException
import java.util.*
import java.util.regex.Pattern

object HashUtil {
    private val HASH_PATTERN = Pattern.compile("[a-zA-Z0-9\\-_]{43}")
    private val UTF8 = StandardCharsets.UTF_8
    fun sha256(input: ByteArray?): String {
        return try {
            val digest = MessageDigest.getInstance("SHA-256")
            digest.update(input)
            val sha256Bytes = digest.digest()
            Base64.getUrlEncoder().withoutPadding().encodeToString(sha256Bytes)
        } catch (ex: NoSuchAlgorithmException) {
            throw RuntimeException(ex)
        }
    }

    @JvmStatic
    fun sha256(input: String): String {
        return sha256(input.toByteArray(UTF8))
    }

    @JvmStatic
    fun isSha256(value: String?): Boolean {
        if (value == null) {
            return false
        }
        return if (value.length != 43) {
            false
        } else HASH_PATTERN.matcher(value).matches()
    }

    fun extractSha256(input: String?): Iterable<String> {
        return object : Iterable<String> {
            override fun iterator(): Iterator<String> {
                return object : Iterator<String> {
                    private val matcher = HASH_PATTERN.matcher(input)
                    private var hasNext = false
                    private var hasNextInitialized = false
                    fun ensureInitialized() {
                        if (!hasNextInitialized) {
                            hasNext = matcher.find()
                            hasNextInitialized = true
                        }
                    }

                    override fun hasNext(): Boolean {
                        ensureInitialized()
                        return hasNext
                    }

                    override fun next(): String {
                        ensureInitialized()
                        hasNextInitialized = false
                        return matcher.group()
                    }
                }
            }
        }
    }

    fun base64encode(input: String): String {
        return Base64.getUrlEncoder().withoutPadding().encodeToString(input.toByteArray(UTF8))
    }

    fun base64decode(input: String): String {
        return String(Base64.getUrlDecoder().decode(input.toByteArray(UTF8)), UTF8)
    }
}